code/game/machinery/dance_machine.dm

```diff
@@ -153,60 +76,120 @@
-				return
-			var/list/available = list()
-			for(var/datum/track/S in songs)
-				available[S.song_name] = S
-			var/selected = params["track"]
-			if(QDELETED(src) || !selected || !istype(available[selected], /datum/track))
-				return
-			selection = available[selected]
+				return TRUE
+
+			var/datum/track/new_song = music_player.songs[params["track"]]
+			if(QDELETED(src) || !istype(new_song, /datum/track))
+				return TRUE
+
+			music_player.selection = new_song
 			return TRUE
+
 		if("set_volume")
 			var/new_volume = params["volume"]
-			if(new_volume == "reset")
-				volume = initial(volume)
-				return TRUE
+			if(new_volume == "reset" || new_volume == "max")
+				music_player.set_volume_to_max()
 			else if(new_volume == "min")
-				volume = 0
-				return TRUE
-			else if(new_volume == "max")
-				volume = initial(volume)
-				return TRUE
-			else if(text2num(new_volume) != null)
-				volume = text2num(new_volume)
-				return TRUE
+				music_player.set_new_volume(0)
+			else if(isnum(text2num(new_volume)))
+				music_player.set_new_volume(text2num(new_volume))
+			return TRUE
+
+		if("loop")
+			music_player.sound_loops = !!params["looping"]
+			return TRUE
 
 /obj/machinery/jukebox/proc/activate_music()
-	active = TRUE
+	if(!isnull(music_player.active_song_sound))
+		return FALSE
+
+	music_player.start_music()
 	update_use_power(ACTIVE_POWER_USE)
 	update_appearance(UPDATE_ICON_STATE)
-	START_PROCESSING(SSobj, src)
-	stop = world.time + selection.song_length
+	if(!music_player.sound_loops)
+		song_timerid = addtimer(CALLBACK(src, PROC_REF(stop_music)), music_player.selection.song_length, TIMER_UNIQUE|TIMER_STOPPABLE|TIMER_DELETE_ME)
+	return TRUE
+
+/obj/machinery/jukebox/proc/stop_music()
+	if(!isnull(song_timerid))
+		deltimer(song_timerid)
+
+	music_player.unlisten_all()
+
+	if(!QDELING(src))
+		COOLDOWN_START(src, jukebox_song_cd, 10 SECONDS)
+		playsound(src,'sound/machines/terminal_off.ogg',50,TRUE)
+		update_use_power(IDLE_POWER_USE)
+		update_appearance(UPDATE_ICON_STATE)
+	return TRUE
+
+/obj/machinery/jukebox/on_set_is_operational(old_value)
+	if(!is_operational)
+		stop_music()
+
+/obj/machinery/jukebox/disco
+	name = "radiant dance machine mark IV"
+	desc = "The first three prototypes were discontinued after mass casualty incidents."
+	icon_state = "disco"
+	base_icon_state = "disco"
+	req_access = list(ACCESS_ENGINEERING)
+	anchored = FALSE
+
+	/// Spotlight effects being played
+	VAR_PRIVATE/list/obj/item/flashlight/spotlight/spotlights = list()
+	/// Sparkle effects being played
+	VAR_PRIVATE/list/obj/effect/overlay/sparkles/sparkles = list()
+
+/obj/machinery/jukebox/disco/indestructible
+	name = "radiant dance machine mark V"
+	desc = "Now redesigned with data gathered from the extensive disco and plasma research."
+	req_access = null
+	anchored = TRUE
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
+	obj_flags = /obj::obj_flags | NO_DECONSTRUCTION
```

Comment:
i saw your other PR, should this be `parent_type`?

Reply:
Yes, it will conflict, I'll change it in this pr